File: gosdk-develop-backup-20260812/gosdk-develop/api/rpcResponse.go
package api

import (
	"fmt"
	"github.com/milon-labs/milon-go-sdk/postcard"
)

type RpcResponseStatus = uint8

const (
	RpcResponseStatusOk          RpcResponseStatus = 0
	RpcResponseStatusInvalid     RpcResponseStatus = 1
	RpcResponseStatusNotFound    RpcResponseStatus = 2
	RpcResponseStatusDisabled    RpcResponseStatus = 3
	RpcResponseStatusUnavailable RpcResponseStatus = 4
	RpcResponseStatusInternal    RpcResponseStatus = 5
	RpcResponseStatusFailed      RpcResponseStatus = 6
)

type RpcResponse struct {
	RequestId uint64            `json:"request_id"`
	Status    RpcResponseStatus `json:"status"`
	Body      []byte            `json:"body"`
	Error     *RpcResponseError `json:"error"`
}

func (rsp *RpcResponse) MarshalPostcard(serializer *postcard.Serializer) error {
	err := serializer.SerializeU64(rsp.RequestId)
	if err != nil {
		return fmt.Errorf("failed to serialize RequestId: %w", err)
	}

	err = serializer.SerializeU8(rsp.Status)
	if err != nil {
		return fmt.Errorf("failed to serialize Status: %w", err)
	}

	err = serializer.SerializeBytes(rsp.Body)
	if err != nil {
		return fmt.Errorf("failed to serialize Body: %w", err)
	}

	err = postcard.SerializeOption(serializer, rsp.Error, func(s *postcard.Serializer, af RpcResponseError) error {
		return af.MarshalPostcard(serializer)
	})
	if err != nil {
		return fmt.Errorf("failed to serialize Error: %w", err)
	}

	return nil
}

func (rsp *RpcResponse) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	var err error

	rsp.RequestId, err = deserializer.DeserializeU64()
	if err != nil {
		return fmt.Errorf("failed to deserialize Message: %w", err)
	}

	rsp.Status, err = deserializer.DeserializeU8()
	if err != nil {
		return fmt.Errorf("failed to deserialize Status: %w", err)
	}

	rsp.Body, err = deserializer.DeserializeBytes()
	if err != nil {
		return fmt.Errorf("failed to deserialize Body: %w", err)
	}
	if rsp.Body == nil {
		rsp.Body = []byte{}
	}

	rsp.Error, err = postcard.DeserializeOption(deserializer, func(d *postcard.Deserializer) (RpcResponseError, error) {
		var af RpcResponseError
		err = af.UnmarshalPostcard(deserializer)
		if err != nil {
			return RpcResponseError{}, fmt.Errorf("failed to deserialize Error: %w", err)
		}
		return af, nil
	})

	return nil
}

type RpcResponseError struct {
	Message string
	Code    *uint16
	Data    *[]byte
}

func (rspErr *RpcResponseError) MarshalPostcard(serializer *postcard.Serializer) error {
	err := serializer.SerializeStr(rspErr.Message)
	if err != nil {
		return fmt.Errorf("failed to serialize Message: %w", err)
	}

	err = postcard.SerializeOption(serializer, rspErr.Code, func(s *postcard.Serializer, code uint16) error {
		return serializer.SerializeU16(code)
	})
	if err != nil {
		return fmt.Errorf("failed to serialize Code: %w", err)
	}

	err = postcard.SerializeOption(serializer, rspErr.Data, func(s *postcard.Serializer, data []byte) error {
		return serializer.SerializeBytes(data)
	})
	if err != nil {
		return fmt.Errorf("failed to serialize Data: %w", err)
	}

	return nil
}

func (rspErr *RpcResponseError) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	var err error

	rspErr.Message, err = deserializer.DeserializeStr()
	if err != nil {
		return fmt.Errorf("failed to deserialize Message: %w", err)
	}

	rspErr.Code, err = postcard.DeserializeOption(deserializer, func(d *postcard.Deserializer) (uint16, error) {
		return deserializer.DeserializeU16()
	})
	if err != nil {
		return fmt.Errorf("failed to deserialize Code: %w", err)
	}

	rspErr.Data, err = postcard.DeserializeOption(deserializer, func(d *postcard.Deserializer) ([]byte, error) {
		return deserializer.DeserializeBytes()
	})
	if err != nil {
		return fmt.Errorf("failed to deserialize Data: %w", err)
	}

	return nil
}

File: gosdk-develop/python/milon_sdk/lib/rpc_response.py
"""RPC 响应线格式（对应 Go lib/rpcResponse.go）。

Postcard 线：
    request_id(u64 变长) + status(u8) + body(bytes) + error(option)
error 内部：message(str) + code(option u16) + data(option bytes)
"""
from __future__ import annotations

from ..postcard import Deserializer, Serializer

RPC_RESPONSE_STATUS_OK = 0
RPC_RESPONSE_STATUS_INVALID = 1
RPC_RESPONSE_STATUS_NOT_FOUND = 2
RPC_RESPONSE_STATUS_DISABLED = 3
RPC_RESPONSE_STATUS_UNAVAILABLE = 4
RPC_RESPONSE_STATUS_INTERNAL = 5
RPC_RESPONSE_STATUS_FAILED = 6


class RpcResponseError:
    __slots__ = ("message", "code", "data")

    def __init__(self, message: str, code: int | None = None, data: bytes | None = None):
        self.message = message
        self.code = code
        self.data = data

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "RpcResponseError":
        message = d.deserialize_str()
        code = d.deserialize_option(lambda dd: dd.deserialize_u16())
        data = d.deserialize_option(lambda dd: dd.deserialize_bytes())
        return cls(message, code, data)

    def __repr__(self) -> str:
        return f"RpcResponseError(message={self.message!r}, code={self.code})"


class RpcResponse:
    __slots__ = ("request_id", "status", "body", "error")

    def __init__(self, request_id: int, status: int, body: bytes, error: RpcResponseError | None = None):
        self.request_id = request_id
        self.status = status
        self.body = bytes(body)
        self.error = error

    def marshal_postcard(self, serializer: Serializer) -> None:
        serializer.serialize_u64(self.request_id)
        serializer.serialize_u8(self.status)
        serializer.serialize_bytes(self.body)
        serializer.serialize_option(self.error is not None, lambda s: self.error.marshal_postcard(s))

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "RpcResponse":
        request_id = d.deserialize_u64()
        status = d.deserialize_u8()
        body = d.deserialize_bytes()
        error = d.deserialize_option(lambda dd: RpcResponseError.unmarshal_postcard(dd))
        return cls(request_id, status, body, error)

    def __repr__(self) -> str:
        return (
            f"RpcResponse(request_id={self.request_id}, status={self.status}, "
            f"body_len={len(self.body)}, error={self.error})"
        )
